README.md
# Created By Facundo Mala

#develop 

[![CircleCI](https://dl.circleci.com/status-badge/img/gh/um-computacion-tm/scrabble-2023-facundomala1/tree/develop.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/um-computacion-tm/scrabble-2023-facundomala1/tree/develop)

#main

[![CircleCI](https://dl.circleci.com/status-badge/img/gh/um-computacion-tm/scrabble-2023-facundomala1/tree/main.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/um-computacion-tm/scrabble-2023-facundomala1/tree/main)

#Maintainability Badge

[![Maintainability](https://api.codeclimate.com/v1/badges/0b3fcb4d1ae1939562e5/maintainability)](https://codeclimate.com/github/um-computacion-tm/scrabble-2023-facundomala1/maintainability)

#Test Coverage Badge

[![Test Coverage](https://api.codeclimate.com/v1/badges/0b3fcb4d1ae1939562e5/test_coverage)](https://codeclimate.com/github/um-computacion-tm/scrabble-2023-facundomala1/test_coverage)



# scrabble-2023-facundomala1

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.1] - 2023-03-05

### Added
Scrabbel game with the following features:

-Date: 31/10/2023

-New functions:

    +The test.sh file was added 

-Changes and fixes:

    +I added changes to  Dockerfile 
    
-Date:23/10/2023

-New functions:---

-Changes and fixes:

    +99% test

-Date: 22/10/2023

-New functions:---

-Changes and fixes:

    +Codes that were extra were deleted
    
-Date: 21/10/2023

-New functions:---

-Changes and fixes:

    +minimal changes in the board class and its tests
    
-Date: 19/10/2023
-New functions:---

-Changes and fixes:

    +The Pyrae function was added to validate the words
    
-Date: 10/10/2023

-New functions:---

-Changes and fixes:

    +Up to 89% of the tests were completed

    
-Date: 08/10/2023

-New functions:---

-Changes and fixes:

    +Fixes in the test code as in game before final delivery
-Date: 05/10/2023

-New functions:---

-Changes and fixes:

    +Fixes in the test code as in game before final delivery

    
-Date: 26/09/2023

-New functions:
    +New class called JokerTile
    +New class TooMuchTilesPut(Exception)
    +New class EmptyTiles(Exception)

-Changes and fixes:

    +Changes to Cell, Board, BagTiles classes

-Date: 25/09/2023

-New functions:--

-Changes and fixes:

    +Add fixes to the Board tests and also to the Cell test to make the tests more precise

    
-Date: 24/09/2023

-New functions:---

-Changes and fixes:

    +Changes in the game code, in the BagTiles and ScrabbleGame classes

-Date: 23/09/2023

-New functions:

    +Add three new functions to the "SC.py" module, to the "Board" class
    +New module to configure Docker

-Changes and fixes: ---

-Date: 12/09/2023

-New functions:

    +Add new functions in "ScrabbleGame" class (draw_game,next_turn,play)

-Changes and fixes:

    +Add new functions in the "Scrabble Game" class (end game)
    +Changes in the "BagTiles" class


-Date: 10/09/2023

-New functions:

    +Add test in "Test Player" to know how many chips the player has

    +New functions for the "Board" class to finish defining the board

-Changes and fixes:

    +Fixes in the "test_init" of the "TestPlayer" class

-Date: 09/09/2023

-New functions:

    +Add other new files (README.py,__init__.py,requirements.txt)

    +Two new tests in TestCell (test_cell_multiplier_both,test_cell_multiplier_word)

-Changes and fixes:---

    
-Date: 08/09/2023

-New functions:

    +Add the "DATA" function which shows you the values of each letter in a better way.

-Changes and fixes:

    +Changes in the "Scrabble Game" class in the "start game" part.

-Date: 06/09/2023

-New functions:

-Changes and fixes:
    +Add changes to the "player" class to summarize and make the code easier to use.

-Date: 29/08/2023

-New functions:---

-Changest and fixes:

    +Add changes 

-Date: 28/08/2023

-New functions:---

-Changes and fixes: 

    +Change the code of the "Bag tiles" to facilitate the counting of chips (some bugs need to be fixed)


-Date: 27/08/2023

-New functions:

    +Add two classes to the game code that are called "Cell" and "Ficha" Cell that has methods for adding a letter, calculating the cell value based on multipliers and multiplier type, and a class called Chip that represents chips with letters and values. 
    
    +Add the code __init__.py

-Changes and fixes: -----


-Date: 24/08/2023

-New functions:

    +Add the new class: Add a new class (player) with its respective test

-Changes and fixes: 

    +changes in the name of a code SC.py

    +fixes in the import in the tests

-Date: 23/08/2023

-New functions:

    +Add the board and token system to find out how much each one is worth

-Changes and fixes: ----
         

- Arabic translation (#444).
- v1.1 French translation.

test.sh
if [ ! -f "/usr/local/bin/codeclimate" ]
then
	curl -L https://github.com/codeclimate/codeclimate/archive/master.tar.gz | tar xvz -C /home/$USER/
	(cd /home/$USER/codeclimate-* && sudo make install)
fi

codeclimate engines:install
python3 tests/test.py
codeclimate analyze 

test/Test.py

import unittest

from GameSC.SC import *

from unittest.mock import patch


class TestTiles(unittest.TestCase):

    def test_tile(self):

        tile = Tile('A', 1)

        self.assertEqual(tile.letter, 'A')

        self.assertEqual(tile.value, 1)


class TestBagTiles(unittest.TestCase):

    @patch('random.shuffle')

    def test_bag_tiles(self, patch_shuffle):

        bag = BagTiles()

        self.assertEqual(

            len(bag.tiles),

            5,

        )

        self.assertEqual(

            patch_shuffle.call_count,

            1,

        )

        self.assertEqual(

            patch_shuffle.call_args[0][0],

            bag.tiles,

        )


    def test_take(self):

        bag = BagTiles()

        tiles = bag.take(2)

        self.assertEqual(

            len(bag.tiles),

            3,

        )

        self.assertEqual(

            len(tiles),

            2,

        )

    def test_put(self):

        bag = BagTiles()

        put_tiles = [Tile('Z', 1), Tile('Y', 1)]

        bag.put(put_tiles)

        self.assertEqual(

            len(bag.tiles),

            7,

        )

class TestPlayer(unittest.TestCase):

    def test_init(self):
        player=Player('Gabriel',0,0,BagTiles())
        self.assertEqual(player.name,'Gabriel')
        self.assertEqual(player.score,0)
        self.assertEqual(player.number,0)
        self.assertEqual(len(player.tiles),0)

    def test_add_tiles(self):
        player=Player('Gabriel',0,0,BagTiles())
        tiles=[Tile('A',1),Tile('B',1),Tile('C',1)]
        player.add_tiles(tiles)
        self.assertEqual(len(player.tiles),3)
        self.assertEqual(player.tiles[0].letter,'A')
        self.assertEqual(player.tiles[1].letter,'B')
        self.assertEqual(player.tiles[2].letter,'C')

    def test_change_tiles(self):
        player=Player('Gabriel',0,0,BagTiles())
        tiles=[Tile('A',1),Tile('B',1),Tile('C',1)]
        player.add_tiles(tiles)
        print(player.tiles)
        player.change_tiles([1,2],[Tile('D',1),Tile('E',1)])
        self.assertEqual(len(player.tiles),3)
        self.assertEqual(player.tiles[0].letter,'D')
        self.assertEqual(player.tiles[1].letter,'E')
        self.assertEqual(player.tiles[2].letter,'C')

class TestBoard(unittest.TestCase):

    def test_init(self):
        board = Board()            
        
        self.assertEqual(len(board.grid),15,)
        
        self.assertEqual(len(board.grid[0]),15,)  

        self.assertEqual(board.grid[0][0].multiplier,3)
        self.assertEqual(board.grid[0][0].multiplier_type,'word')
        self.assertEqual(board.grid[0][0].letter,None)      

    

class TestCell(unittest.TestCase):

    def test_init(self):
        cell = Cell(multiplier=2, multiplier_type='letter')
        self.assertEqual(cell.multiplier, 2)
        self.assertEqual(cell.multiplier_type, 'letter')
        self.assertIsNone(cell.letter)
        self.assertEqual(cell.calculate_value(), 0)

    def test_add_letter(self):
        cell = Cell(multiplier=1, multiplier_type='')
        letter = Tile(letter='p', value=3)
        cell.add_letter(letter=letter)
        self.assertEqual(cell.letter, letter)

    def test_cell_value(self):
        cell = Cell(multiplier=2, multiplier_type='letter')
        letter = Tile(letter='p', value=3)
        cell.add_letter(letter=letter)
        self.assertEqual(cell.calculate_value(), 6)

    def test_cell_multiplier_word(self):
        cell_1 = Cell(multiplier=1,multiplier_type='letter')
        cell_1.add_letter(Tile('C', 1)) 
        cell_2 = Cell(multiplier=1,multiplier_type='letter')
        cell_2.add_letter(Tile('A', 1))
        cell_3 = Cell(multiplier=3,multiplier_type='word')
        cell_3.add_letter(Tile('S', 2))
        cell_4 = Cell(multiplier=1,multiplier_type='letter')
        cell_4.add_letter(Tile('A', 1))
        word = [cell_1, cell_2, cell_3, cell_4]
        value=Board().calculate_word_value(word)
        self.assertEqual(value,15)

    def test_cell_multiplier_both(self):
        cell_1 = Cell(multiplier=2,multiplier_type='letter')
        cell_1.add_letter(Tile('C', 1)) 
        cell_2 = Cell(multiplier=1,multiplier_type='letter')
        cell_2.add_letter(Tile('A', 1))
        cell_3 = Cell(multiplier=3,multiplier_type='word')
        cell_3.add_letter(Tile('S', 2))
        cell_4 = Cell(multiplier=1,multiplier_type='letter')
        cell_4.add_letter(Tile('A', 1))
        word = [cell_1, cell_2, cell_3, cell_4]
        value=Board().calculate_word_value(word)
        self.assertEqual(value,18)





if __name__ == '__main__':

    unittest.main()



test/Testing.py

import unittest

from GameSC.SC import Tile, BagTiles, Player, Board, Cell



from unittest.mock import patch


class TestTiles(unittest.TestCase):

    def test_tile(self):

        tile = Tile('A', 1)

        self.assertEqual(tile.letter, 'A')

        self.assertEqual(tile.value, 1)


class TestBagTiles(unittest.TestCase):

    @patch('random.shuffle')

    def test_bag_tiles(self, patch_shuffle):

        bag = BagTiles()

        self.assertEqual(

            len(bag.tiles),

            5,

        )

        self.assertEqual(

            patch_shuffle.call_count,

            1,

        )

        self.assertEqual(

            patch_shuffle.call_args[0][0],

            bag.tiles,

        )


    def test_take(self):

        bag = BagTiles()

        tiles = bag.take(2)

        self.assertEqual(

            len(bag.tiles),

            3,

        )

        self.assertEqual(

            len(tiles),

            2,

        )

    def test_put(self):

        bag = BagTiles()

        put_tiles = [Tile('Z', 1), Tile('Y', 1)]

        bag.put(put_tiles)

        self.assertEqual(

            len(bag.tiles),

            7,

        )

class TestPlayer(unittest.TestCase):

    def test_init(self):
        player=Player('Gabriel',0,0,BagTiles())
        self.assertEqual(player.name,'Gabriel')
        self.assertEqual(player.score,0)
        self.assertEqual(player.number,0)
        self.assertEqual(len(player.tiles),0)

    def test_add_tiles(self):
        player=Player('Gabriel',0,0,BagTiles())
        tiles=[Tile('A',1),Tile('B',1),Tile('C',1)]
        player.add_tiles(tiles)
        self.assertEqual(len(player.tiles),3)
        self.assertEqual(player.tiles[0].letter,'A')
        self.assertEqual(player.tiles[1].letter,'B')
        self.assertEqual(player.tiles[2].letter,'C')

    def test_change_tiles(self):
        player=Player('Gabriel',0,0,BagTiles())
        tiles=[Tile('A',1),Tile('B',1),Tile('C',1)]
        player.add_tiles(tiles)
        print(player.tiles)
        player.change_tiles([1,2],[Tile('D',1),Tile('E',1)])
        self.assertEqual(len(player.tiles),3)
        self.assertEqual(player.tiles[0].letter,'D')
        self.assertEqual(player.tiles[1].letter,'E')
        self.assertEqual(player.tiles[2].letter,'C')

class TestBoard(unittest.TestCase):

    def test_init(self):
        board = Board()            
        
        self.assertEqual(len(board.grid),15,)
        
        self.assertEqual(len(board.grid[0]),15,)  

        self.assertEqual(board.grid[0][0].multiplier,3)
        self.assertEqual(board.grid[0][0].multiplier_type,'word')
        self.assertEqual(board.grid[0][0].letter,None)      

    

class TestCell(unittest.TestCase):

    def test_init(self):
        cell = Cell(multiplier=2, multiplier_type='letter')
        self.assertEqual(cell.multiplier, 2)
        self.assertEqual(cell.multiplier_type, 'letter')
        self.assertIsNone(cell.letter)
        self.assertEqual(cell.calculate_value(), 0)

    def test_add_letter(self):
        cell = Cell(multiplier=1, multiplier_type='')
        letter = Tile(letter='p', value=3)
        cell.add_letter(letter=letter)
        self.assertEqual(cell.letter, letter)

    def test_cell_value(self):
        cell = Cell(multiplier=2, multiplier_type='letter')
        letter = Tile(letter='p', value=3)
        cell.add_letter(letter=letter)
        self.assertEqual(cell.calculate_value(), 6)

    def test_cell_multiplier_word(self):
        cell_1 = Cell(multiplier=1,multiplier_type='letter')
        cell_1.add_letter(Tile('C', 1)) 
        cell_2 = Cell(multiplier=1,multiplier_type='letter')
        cell_2.add_letter(Tile('A', 1))
        cell_3 = Cell(multiplier=3,multiplier_type='word')
        cell_3.add_letter(Tile('S', 2))
        cell_4 = Cell(multiplier=1,multiplier_type='letter')
        cell_4.add_letter(Tile('A', 1))
        word = [cell_1, cell_2, cell_3, cell_4]
        value=Board().calculate_word_value(word)
        self.assertEqual(value,15)

    def test_cell_multiplier_both(self):
        cell_1 = Cell(multiplier=2,multiplier_type='letter')
        cell_1.add_letter(Tile('C', 1)) 
        cell_2 = Cell(multiplier=1,multiplier_type='letter')
        cell_2.add_letter(Tile('A', 1))
        cell_3 = Cell(multiplier=3,multiplier_type='word')
        cell_3.add_letter(Tile('S', 2))
        cell_4 = Cell(multiplier=1,multiplier_type='letter')
        cell_4.add_letter(Tile('A', 1))
        word = [cell_1, cell_2, cell_3, cell_4]
        value=Board().calculate_word_value(word)
        self.assertEqual(value,18)





if __name__ == '__main__':

    unittest.main()



game/SC.py
import random 

TOTALTILES = 100

class TooMuchTilesPut(Exception):
    pass

class EmptyTiles(Exception):
    pass

class MuchTilesPut(Exception):
    pass

class MuchTiles(Exception):
    pass

class TooMuchTiles (Exception):
    pass

class ScrabbleGame:

    def __init__(self, players_count):

        self.players = []
        self.bag_tiles = BagTiles(fichas_disponibles)
        self.current_player = None
        self.board = Board()
        for number in range(players_count):
            self.players.append(Player(number=number))

    def start_game(self):
            
            for player in self.players:
                player.add_tiles(self.bag_tiles.take(7))
            self.current_player = self.players[0]   

    def next_turn(self):
        if self.current_player == None:
            self.current_player = self.players[0]
        elif self.current_player == self.players[-1]:
            self.current_player = self.players[0]
        else:
            index=self.players.index(self.current_player)+1
            self.current_player=self.players[index]


    def validate_word(self, word):
        if word == "":
            return True
        else:
            return False

    def end_game(self):

        if  self.bag_tiles == []:
            return True
        else:
            return False
        
        


class Tile:

    def __init__(self, letter, value):

        self.letter = letter
        self.value = value

class JokerTile(Tile):

    def __init__(self, letter, value):
        super().__init__(letter, value)

    def chooseLetter(self,letter_joker):
        for i in DATA:
            if i['letter']==letter_joker.upper():
                self.letter=letter_joker.upper()
                self.value=i['value']
                break
        else:
            raise EmptyTiles         
    
DATA= [
    {"letter": "A", "value": 1, "quantity": 12},
    {"letter": "B", "value": 3, "quantity": 2},
    {"letter": "C", "value": 3, "quantity": 4},
    {"letter": "D", "value": 2, "quantity": 5},
    {"letter": "E", "value": 1, "quantity": 12},
    {"letter": "F", "value": 4, "quantity": 1},
    {"letter": "G", "value": 2, "quantity": 2},
    {"letter": "H", "value": 4, "quantity": 2},
    {"letter": "I", "value": 1, "quantity": 6},
    {"letter": "J", "value": 8, "quantity": 1},
    {"letter": "L", "value": 1, "quantity": 4},
    {"letter": "M", "value": 3, "quantity": 2},
    {"letter": "N", "value": 1, "quantity": 5},
    {"letter": "Ñ", "value": 8, "quantity": 1},
    {"letter": "O", "value": 1, "quantity": 9},
    {"letter": "P", "value": 3, "quantity": 2},
    {"letter": "Q", "value": 5, "quantity": 1},
    {"letter": "R", "value": 1, "quantity": 5},
    {"letter": "S", "value": 1, "quantity": 6},
    {"letter": "T", "value": 1, "quantity": 4},
    {"letter": "U", "value": 1, "quantity": 5},
    {"letter": "V", "value": 4, "quantity": 1},
    {"letter": "X", "value": 8, "quantity": 1},
    {"letter": "Y", "value": 4, "quantity": 1},
    {"letter": "Z", "value": 10, "quantity": 1},
    {"letter": "CH", "value": 5, "quantity": 1},
    {"letter": "LL", "value": 8, "quantity": 1},
    {"letter": "RR", "value": 8, "quantity": 1},
    {"letter": "_", "value": 0, "quantity": 2} ]
   

class BagTiles:
    def __init__(self):
        self.tiles=[]
        for i in DATA:
            for j in range(i.get('quantity')):
                self.tiles.append(Tile(i.get("letter"),i.get("value")))
        random.shuffle(self.tiles)
 
    def draw_tiles(self,cantidad):
        tile_drawn=[]
        try:
            if cantidad>len(self.tiles):
                raise TooMuchTiles
            else:
                for i in range(cantidad):
                    tile_drawn.append(self.tiles.pop())
                return tile_drawn
        except TooMuchTiles:
            return tile_drawn


    def put_tiles(self,tiles:list):
        try:
            if len(tiles)+len(self.tiles)<=TOTALTILES:
                self.tiles.extend(tiles)
            else:
                raise TooMuchTilesPut
        except TooMuchTilesPut:
            return TooMuchTilesPut

    def tiles_remaining(self):
        return len(self.tiles)

# Definición de las fichas disponibles
fichas_disponibles = [
    Tile('A', 1),
    Tile('A', 1),
    Tile('A', 1),
    Tile('A', 1),
    Tile('A', 1),
    #(otras fichas con sus letras y valores)
]

# Crear una instancia de BagTiles con las fichas disponibles
bag = BagTiles(fichas_disponibles)

# Definir la función de selección de fichas
def seleccionar_fichas(num_fichas):
    return bag.take(num_fichas)

# Función para calcular la puntuación de una palabra
def calcular_puntuacion(palabra, casillas_usadas):
    puntuacion = 0
    multiplicador = 1

    for i, letra in enumerate(palabra):
        if casillas_usadas[i] == "DL":
            puntuacion += Tile[letra][1] * 2
        elif casillas_usadas[i] == "TL":
            puntuacion += Tile[letra][1] * 3
        elif casillas_usadas[i] == "DW":
            multiplicador *= 2
            puntuacion += Tile[letra][1]
        elif casillas_usadas[i] == "TW":
            multiplicador *= 3
            puntuacion += Tile[letra][1]
        else:
            puntuacion += Tile[letra][1]

    return puntuacion * multiplicador

#tablero
class Board:
    #revisar
    def __init__(self,grid=None):
        self.grid = [[ Cell(1, '') for _ in range(15) ]for _ in range(15)]
        self.grid[7][7].multiplier_type = 'word'
        self.grid[7][7].multiplier = 3
        self.grid[0][0].multiplier_type = 'word'
        self.grid[0][0].multiplier = 3
        self.grid[0][7].multiplier_type = 'word'

    def calculate_word_value(self, word):
        value = 0
        for cell in word:
            value += cell.calculate_value()
        for cell in word:
            if cell.multiplier_type == 'word':
                value *= cell.multiplier
                cell.multiplier = 1
        return value
    
    def validate_len_of_word_in_board(self, word, location, orientation):
        location_x = location[0]
        location_y = location[1]
        len_word = len(word)
        if orientation == 'H':
            if location_x + len_word > 15:
                return False
            else:
                return True
        else:
            if location_y + len_word > 15:
                return False
            else:
                return True
        
    def put_word(self,word,location, orientation):
        location_x = location[0]
        location_y = location[1]
        if orientation == 'V':
            for i in range(len(word)):
                self.grid[location_x+i][location_y].add_letter(word[i])
        else:
            for i in range(len(word)):
                self.grid[location_x][location_y+i].add_letter(word[i])

    def is_empty(self):
        if self.grid[7][7].letter is None:
            return True


    tablero = [
        ["TW", "", "", "DL", "", "", "", "TW", "", "", "", "DL", "", "", "TW"],
        ["", "DW", "", "", "", "TL", "", "", "", "TL", "", "", "", "DW", ""],
        ["", "", "DW", "", "", "", "DL", "", "DL", "", "", "", "DW", "", ""],
        ["DL", "", "", "DW", "", "", "", "DL", "", "", "", "DW", "", "", "DL"],
        ["", "", "", "", "DW", "", "", "", "", "", "DW", "", "", "", ""],
        ["", "TL", "", "", "", "TL", "", "", "", "TL", "", "", "", "TL", ""],
        ["", "", "DL", "", "", "", "DL", "", "DL", "", "", "", "DL", "", ""],
        ["TW", "", "", "DL", "", "", "", "S", "", "", "", "DL", "", "", "TW"],
        ["", "", "DL", "", "", "", "DL", "", "DL", "", "", "", "DL", "", ""],
        ["", "TL", "", "", "", "TL", "", "", "", "TL", "", "", "", "TL", ""],
        ["", "", "", "", "DW", "", "", "", "", "", "DW", "", "", "", ""],
        ["DL", "", "", "DW", "", "", "", "DL", "", "", "", "DW", "", "", "DL"],
        ["", "", "DW", "", "", "", "DL", "", "DL", "", "", "", "DW", "", ""],
        ["", "DW", "", "", "", "TL", "", "", "", "TL", "", "", "", "DW", ""],
        ["TW", "", "", "DL", "", "", "", "TW", "", "", "", "DL", "", "", "TW"]
    ]

    

class Cell:

    def __init__(self, multiplier, multiplier_type='',letter=None,active=True):

        self.multiplier = multiplier

        self.multiplier_type = multiplier_type

        self.letter = None


    def add_letter(self,tile):

        self.letter = tile


    def calculate_value(self):

        if self.letter is None:

            return 0

        if self.multiplier_type == 'letter':

            return self.letter.value * self.multiplier

        else:

            return self.letter.value


class Player:
    
        def __init__(self,name="",number=0,score=0,bag_tiles=0):
            self.name = name
            self.score = score
            self.bag_tiles = bag_tiles
            self.number = number
            self.tiles = []

        
        def add_tiles(self,tiles):
            self.tiles.extend(tiles)

        def remove_tiles(self,tiles):
            for tile in tiles:
                self.tiles.remove(tile)
        
        def change_tiles(self,player_old_tiles_index=[],player_new_tiles=[]):
            tiles_to_change=[]
            for tile_index in range (len(player_old_tiles_index)):
                tiles_to_change.append(self.tiles[player_old_tiles_index[tile_index]-1])
                self.tiles[player_old_tiles_index[tile_index]-1]=player_new_tiles[player_old_tiles_index[tile_index]-1]
            return tiles_to_change
    


game/game_scrabble.py
import random 

from pyrae import dle


TOTALTILES = 100

class TooMuchTilesPut(Exception):
    pass

class EmptyTiles(Exception):
    pass

class MuchTilesPut(Exception):
    pass

class MuchTiles(Exception):
    pass

class TooMuchTiles (Exception):
    pass


class ScrabbleGame:

    def __init__(self, players_count: int):
        self.board = Board()
        self.bag_tiles = BagTiles()
        self.players:list[Player] = []
        for _ in range(players_count):
            self.players.append(Player())
        self.current_player = None
        self.turn = 0

    def next_turn(self):
        if self.current_player == None:
            self.current_player = self.players[0]
        elif self.current_player == self.players[-1]:
            self.current_player = self.players[0]
        else:
            index=self.players.index(self.current_player)+1
            self.current_player=self.players[index]
            
    def distribute_tiles(self):
        for player in self.players:
            player.add_tiles(self.bag_tiles.draw_tiles(7))

    def validate_word(self, word):
        return self.board.validate_word(word)

    def show_board(self):
        return self.board.show_board()

    def show_player_tiles(self):
        return self.current_player.show_tiles()

    def end_game(self):
        if  self.bag_tiles == []:
            return True
        else:
            return False
        
        
        


class Tile:

    def __init__(self, letter, value):

        self.letter = letter
        self.value = value

 
    
DATA= [
    {"letter": "A", "value": 1, "quantity": 12},
    {"letter": "B", "value": 3, "quantity": 2},
    {"letter": "C", "value": 3, "quantity": 4},
    {"letter": "D", "value": 2, "quantity": 5},
    {"letter": "E", "value": 1, "quantity": 12},
    {"letter": "F", "value": 4, "quantity": 1},
    {"letter": "G", "value": 2, "quantity": 2},
    {"letter": "H", "value": 4, "quantity": 2},
    {"letter": "I", "value": 1, "quantity": 6},
    {"letter": "J", "value": 8, "quantity": 1},
    {"letter": "L", "value": 1, "quantity": 4},
    {"letter": "M", "value": 3, "quantity": 2},
    {"letter": "N", "value": 1, "quantity": 5},
    {"letter": "Ñ", "value": 8, "quantity": 1},
    {"letter": "O", "value": 1, "quantity": 9},
    {"letter": "P", "value": 3, "quantity": 2},
    {"letter": "Q", "value": 5, "quantity": 1},
    {"letter": "R", "value": 1, "quantity": 5},
    {"letter": "S", "value": 1, "quantity": 6},
    {"letter": "T", "value": 1, "quantity": 4},
    {"letter": "U", "value": 1, "quantity": 5},
    {"letter": "V", "value": 4, "quantity": 1},
    {"letter": "X", "value": 8, "quantity": 1},
    {"letter": "Y", "value": 4, "quantity": 1},
    {"letter": "Z", "value": 10, "quantity": 1},
    {"letter": "CH", "value": 5, "quantity": 1},
    {"letter": "LL", "value": 8, "quantity": 1},
    {"letter": "RR", "value": 8, "quantity": 1},
    {"letter": "_", "value": 0, "quantity": 2} ]
   

class JokerTile(Tile):
    
    def __init__(self, letter, value):
        super().__init__(letter, value)


    def chooseLetter(self,letter_joker):
        for i in DATA:
            if i['letter']==letter_joker.upper():
                self.letter=letter_joker.upper()
                self.value=0
                break
        else:
            raise EmptyTiles 
        
class BagTiles:
    def __init__(self):
        self.tiles=[]
        for i in DATA:
            for j in range(i.get('quantity')):
                self.tiles.append(Tile(i.get("letter"),i.get("value")))
        random.shuffle(self.tiles)
 
    def draw_tiles(self,cantidad):
        tile_drawn=[]
        try:
            if cantidad>len(self.tiles):
                raise TooMuchTiles
            else:
                for i in range(cantidad):
                    tile_drawn.append(self.tiles.pop())
                return tile_drawn
        except TooMuchTiles:
            return tile_drawn


    def put_tiles(self,tiles:list):
        try:
            if len(tiles)+len(self.tiles)<=TOTALTILES:
                self.tiles.extend(tiles)
            else:
                raise TooMuchTilesPut
        except TooMuchTilesPut:
            return TooMuchTilesPut

    def tiles_remaining(self):
        return len(self.tiles)




#tablero
class Board:
    def __init__(self):
        self.grid = [
            [ Cell(None,None,1, '') for _ in range(15) ]
            for _ in range(15)
        ]

    #def calculate_word_value(self, word):
        #value = 0
        #for cell in word:
         #   value += cell.calculate_value()
        #for cell in word:
         #   if cell.multiplier_type == 'word':
          #      value *= cell.multiplier
           #     cell.multiplier = 1
        #return value

    
    def validate_word_inside_board(self, word, location, orientation):
        self.word = word
        self.orientation = orientation
        self.position_row = location[0]
        self.position_col = location[1] 
        if orientation == 'H' and len(self.word)<=15-self.position_col:
            return True
        elif orientation == 'V' and len(self.word)<=15-self.position_row:
            return True
        else:
            return False
        
    def empty(self):
        if self.grid[7][7].letter == None:
            self.is_empty = True
        else:
            self.is_empty = False
   
    def validate_word_place_board(self, word, location, orientation):
        valid = self.validate_word_inside_board(word, location, orientation)
        self.empty()
        if self.is_empty == False:
            if valid == True:
                if orientation == "H":
                    for i in self.word:
                        index = self.position_col
                        if self.grid[self.position_row][index].letter is not None:
                            if i != self.grid[self.position_row][index].letter.letter:
                                return False
                        self.position_col += 1
                    return True
                else:
                    for i in self.word:
                        index = self.position_row
                        if self.grid[index][self.position_col].letter is not None:
                            if i != self.grid[index][self.position_col].letter.letter:    
                                return False
                        self.position_row += 1
                    return True
        else:
            if valid == True:
                for i in self.word:
                    if orientation == "H":
                        index = self.position_col
                        if self.position_row == 7 and index == 7:
                            return True
                        self.position_col += 1 
                    else:
                        index = self.position_row
                        if self.position_col == 7 and index == 7:
                            return True
                        self.position_row += 1 
                return False
    
    def validate_word(self, word):

        flag=dle.search_by_word(word)
        if word.lower() not in flag.title:
            return False
        else:
            return True
        
    def show_board(self):
        print('')
        columnas = ['   ','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O']
        print("  ".join(columnas))
        print('-------------------------------------------------')
        filas  = ['01','02','03','04','05','06','07','08','09','10','11','12','13','14','15']
        for i in range(15):
            print(filas[i], end='|  ')
            for j in range(15):
                if self.grid[i][j].letter is None:
                    print('-', end='  ')
                else:
                    print(self.grid[i][j].letter.letter.upper(), end='  ')
            print('')
        print('')   
    
    

    

class Cell:

    def __init__(self,letter,state,multiplier, multiplier_type):
        self.multiplier = multiplier
        self.multiplier_type = multiplier_type
        self.letter=letter
        self.state=state
        
        
    def add_letter(self, letter:Tile):
        self.letter = letter

    def calculate_value(self):
        if self.letter is None:
            return 0
        if self.multiplier_type == 'letter'and self.state==True:
            return self.letter.value * self.multiplier
        else:
            return self.letter.value
        
    

class Player:
    
        def __init__(self,name="",number=0,score=0,bag_tiles=0):
            self.name = name
            self.score = score
            self.bag_tiles = bag_tiles
            self.number = number
            self.tiles = []

        
        def add_tiles(self,tiles):
            self.tiles.extend(tiles)

        def change_tiles(self,player_old_tiles_index=[],player_new_tiles=[]):
            tiles_to_change=[]
            for tile_index in range (len(player_old_tiles_index)):
                tiles_to_change.append(self.tiles[player_old_tiles_index[tile_index]-1])
                self.tiles[player_old_tiles_index[tile_index]-1]=player_new_tiles[player_old_tiles_index[tile_index]-1]
            return tiles_to_change
    
        def show_tiles(self):
            return self.tiles